### PR TITLE
Netatmo: Add api key provider and custom api plugin settings

### PR DESCRIPTION
**Pull request by Boernsman**
nymea-plugins pull request checklist:

- [x] Make sure the pull request's title is of format "Plugin name: Add support for xyz" or "New plugin: Plugin name"

- [x] Did you test the changes on hardware, if not (e.g. absence of required hardware), please mention a person to confirm it has been tested.

- [x] Did you update the plugin's README.md accordingly?

- [x] Did you update translations (`cd builddir && make lupdate`)?

- [x] If you added a new plugin, should it be added to nymea-plugins-all or nymea-plugins-maker?

### CODE REVIEW

File: netatmo/integrationpluginnetatmo.cpp [hunk 1]
```diff
@@ -42,17 +42,33 @@ IntegrationPluginNetatmo::IntegrationPluginNetatmo()
...
+    connect(apiKeyStorage(), &ApiKeyStorage::keyAdded, this, &IntegrationPluginNetatmo::updateClientCredentials);
+    connect(apiKeyStorage(), &ApiKeyStorage::keyUpdated, this, &IntegrationPluginNetatmo::updateClientCredentials);
+}
+
 void IntegrationPluginNetatmo::startPairing(ThingPairingInfo *info)
 {
+    // Checking the client credentials
+    if (m_clientId.isEmpty() || m_clientSecret.isEmpty()) {
+        qCWarning(dcNetatmo()) << "Pairing failed, client credentials are not set";
+        info->finish(Thing::ThingErrorAuthenticationFailure, QT_TR_NOOP("Client credentials are not set."));
```
Comment (mzanetti) [inline netatmo/integrationpluginnetatmo.cpp:59]: I personally find this message a bit misleading... If I read this I'd expect something wrong with the stuff I entered but not with the setup. Dunno, maybe it's just me. Just brainstorming: From an end user point of view, what would be the thing that makes you understand what's actually going on? Something like "No API key installed" perhaps? I don't like API key in a end user context as it seems too technical, client credentials seems to unspecific tho to know what to look for... I think the "not installed" part would be good tho as it lets the user know that something needs installing?

Dunno... I'll leave it up to you... just some general thoughts.

**Review by mzanetti — CHANGES_REQUESTED** [review]
I believe there's on issue to be fixed, at least to be verified.

File: netatmo/integrationpluginnetatmo.cpp [hunk 2]
```diff
@@ -61,34 +77,35 @@ void IntegrationPluginNetatmo::startPairing(ThingPairingInfo *info)
...
+            m_authentications.insert(authentication, info->thingId());
             info->finish(Thing::ThingErrorNoError);
         } else {
+            qCWarning(dcNetatmo()) << "Wrong username or password";
             info->finish(Thing::ThingErrorAuthenticationFailure, QT_TR_NOOP("Wrong username or password"));
         }
     });
     authentication->startAuthentication();
-    connect(info, &QObject::destroyed, authentication, &OAuth2::deleteLater);
+    connect(info, &ThingPairingInfo::aborted, authentication, &OAuth2::deleteLater);
```
[the remark this answers is not in the document]
Comment (Boernsman) [reply netatmo/integrationpluginnetatmo.cpp:100]: If the authentication was successfull the authentication object will now be forwareded to the thing setup. It is intended to delete the object only on failure

**Review by mzanetti — APPROVED** [review]

File: netatmo/integrationpluginnetatmo.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (Boernsman) [reply netatmo/integrationpluginnetatmo.cpp:100]: reworked the reconfiguration, no all memory leaks should be fixed

File: netatmo/integrationpluginnetatmo.cpp (hunk 1 above)
Comment (Boernsman) [reply netatmo/integrationpluginnetatmo.cpp]: error message changed to "No API key installed"

**Pull request merged**